app/app.go

```diff
@@ -0,0 +1,122 @@
+	Headers    http.Header
+	StatusCode int
+	Body       []byte
+}
+
+// Config is the app configuration used in a Provider for instantiating a new App.
+// It contains kubernetes configuration for communicating with an API server, the App's ManifestData as fetched
+// by the runner, and additional arbitrary configuration details that may be app-specific.
+type Config struct {
+	// KubeConfig is a kubernetes rest.Config used to communicate with the API server where the App's Kinds are stored.
+	KubeConfig rest.Config
+	// ManifestData is the fetched ManifestData the runner is using for determining app kinds and capabilities.
+	ManifestData ManifestData
+	// SpecificConfig is app-specific config (as opposed to generic config)
+	SpecificConfig SpecificConfig
+}
+
+// SpecificConfig is app-specific configuration which can vary from app to app
+// TODO: better type than any
+type SpecificConfig any
+
+// Provider represents a type which can provide an app manifest, and create a new App when given a configuration.
+// It should be used by runners to determine an app's capabilities and create an instance of the app to run.
+type Provider interface {
+	// Manifest returns a Manifest, which may contain ManifestData or may point to a location where ManifestData can be fetched from.
+	// The runner should use the ManifestData to determine app capabilities.
+	Manifest() Manifest
+	// SpecificConfig is any app-specific config that cannot be loaded by the runner that should be provided in NewApp
+	SpecificConfig() SpecificConfig
+	// NewApp creates a new App instance using the provided config, or returns an error if an App cannot be instantiated.
+	NewApp(Config) (App, error)
+}
+
+// Runnable represents a type which can be run until it errors or the provided channel is stopped (or receives a message)
+type Runnable interface {
+	// Run runs the process and blocks until one of the following conditions are met:
+	// * An unrecoverable error occurs, in which case it returns the error
+	// * The provided context completes
+	// * The process completes and does not need to run again
+	Run(context.Context) error
+}
+
+type AdmissionRequest resource.AdmissionRequest
+type MutatingResponse resource.MutatingResponse
+
+// App represents an app platform application logical structure.
+// An App is typically run with a wrapper, such as simple.NewStandaloneOperator,
+// which will present a runtime layer (such as kubernetes webhooks in the case of an operator),
+// and translate those into calls to the App. The wrapper is typically also responsible for lifecycle management
+// and running the Runnable provided by Runner().
+// Pre-built implementations of App exist in the simple package, but any type which implements App
+// should be capable of being run by an app wrapper.
+type App interface {
+	// Validate validates the incoming request, and returns an error if validation fails
+	Validate(ctx context.Context, request *resource.AdmissionRequest) error
```

Comment:
very nit: since we have `AdmissionRequest` and `MutatingResponse` re-declared above, should we use them?

Reply:
Yes, we should be using them here, I completely forgot to do so 🤦 . Good catch.

Reply:
OK, implementing was a bit more work than I anticipated, but `Validate` and `Mutate` now use the `app.AdmissionRequest` and `app.MutatingResponse`. Had to introduce some new interfaces and types in `simple` (`simple.KindValidator` and its simple implementation `simple.Validator`, and `simple.KindMutator` and its implementation `simple.Mutator`), and then a translation in `operator.Runner` because `k8s.WebhookServer` wants the `resource` variants. Probably future work to make this whole chain more efficient, but for now, the API for `app.App` is correct, which is what matters more.

Reply:
Nice, thank you!